lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/grok/Discovery.java

```diff
@@ -0,0 +1,196 @@
+
+        for (Grok g : groky) {
+            mGrok.put(g.getSaved_pattern(), g);
+        }
+        return mGrok;
+    }
+
+    /**
+     *
+     * @param expandedPattern regex string
+     * @return the complexity of the regex
+     */
+    private int complexity(String expandedPattern)
+    {
+        int score = 0;
+
+        score += expandedPattern.split("\\Q" + "|" + "\\E", -1).length - 1;
+        score += expandedPattern.length();
+
+        return score;
+    }
+
+    /**
+     * Find a pattern from a log.
+     *
+     * @param text witch is the representation of your single
+     * @return Grok pattern %{Foo}...
+     */
+    public String discover(String text)
+            throws GrokException
+    {
+        if (text == null) {
+            return "";
+        }
+
+        Map<String, Grok> groks = new TreeMap<String, Grok>();
+        Map<String, String> gPatterns = grok.getPatterns();
+        String texte = text;
+
+        // Compile the pattern
+        Iterator<Entry<String, String>> it = gPatterns.entrySet().iterator();
+        while (it.hasNext()) {
+            @SuppressWarnings("rawtypes")
+            Map.Entry pairs = (Map.Entry) it.next();
+            String key = pairs.getKey().toString();
+            Grok g = new Grok();
+
+            try {
+                g.copyPatterns(gPatterns);
+                g.setSaved_pattern(key);
+                g.compile("%{" + key + "}");
+                groks.put(key, g);
+            }
+            catch (GrokException e) {
+                // TODO: Add logger
+                continue;
+            }
+        }
+
+        // Sort patterns by complexity
+        Map<String, Grok> patterns = this.sort(groks);
+
+        Iterator<Entry<String, Grok>> pit = patterns.entrySet().iterator();
+        while (pit.hasNext()) {
+            @SuppressWarnings("rawtypes")
+            Map.Entry pairs = (Map.Entry) pit.next();
+            String key = pairs.getKey().toString();
+            Grok value = (Grok) pairs.getValue();
+
+            // We want to search with more complex pattern
+            // We avoid word, small number, space....
+            if (this.complexity(value.getNamedRegex()) < 20) {
+                continue;
+            }
+
+            Match m = value.match(text);
+            if (m.isNull()) {
+                continue;
+            }
+            // get the part of the matched text
+            String part = getPart(m, text);
+
+            // we skip boundary word
+            Pattern pattern = Pattern.compile(".\\b.");
+            Matcher ma = pattern.matcher(part);
+            if (!ma.find()) {
+                continue;
+            }
+
+            // We skip the part that already include %{Foo}
+            Pattern pattern2 = Pattern.compile("%\\{[^}+]\\}");
+            Matcher ma2 = pattern2.matcher(part);
+
+            if (ma2.find()) {
+                continue;
+            }
+            texte = StringUtils.replace(texte, part, "%{" + key + "}");
```

Comment:
I don't think we need to pull in all of commons lang for this function.  I think this is the equivilent in moder java (but please verify):
```
texte = texte.replace(part, "%{" + key + "}");
```